[v2] overlay2: test for and report metacopy status

neersighted opened the pull request:
This is helpful for debugging issues with the overlay2 storage driver.

Metacopy may be enabled by CONFIG_OVERLAY_FS_METACOPY, the overlay.metacopy parameter, or the metacopy mount option. See Documentation/filesystems/overlayfs.txt in the kernel source tree for details.

Backported from containers/storage@05c69f1b2a.

neersighted committed:
overlay2: test for and report metacopy status

This is a first, naive implementation, that does not account for
userxattr/UserNS.

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

thaJeztah commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -159,6 +160,18 @@ func Init(home string, options []string, idMap idtools.IdentityMapping) (graphdr
 
+	usingMetacopy, err := doesMetacopy(testdir)
+	if err == nil {
+		if usingMetacopy {
+			logger.Debugf("overlay test mount indicated that metacopy is being used")
+		} else {
+			logger.Debugf("overlay test mount indicated that metacopy is not being used")
+		}
```
Perhaps make this a single log where we print the value (could be using `WithField()` for that);

```go
logger.WithField("usingMetacopy", usingMetacopy).Debug("successfully detected meta-copy support")
```

neersighted replied:
Ack

thaJeztah reviewed (changes requested):
Thanks! Had a first glance over the changes; out of interest; this currently only does the detection; should (if supported) this detection also be used by the actual driver?

neersighted commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -107,3 +107,59 @@ func doesSupportNativeDiff(d string) error {
+// metadata by using nodes marked with an "overlay.metacopy" attribute to avoid
+// copying up a file from a lower layer unless/until its contents are being
+// modified
+func doesMetacopy(d string) (bool, error) {
```
The important thing to note here is that we are testing if metacopy is being used -- we (the daemon) don't actually ask for it to be enabled, or not enabled. That is left entirely up to the host OS (which may have it set due to `CONFIG_OVERLAY_FS_METACOPY` or `overlay2.metacopy`).

This is meant to indicate if metacopy is being used, not if it is supported. Obviously if it is unsupported it will not be used.

thaJeztah replied:
> The important thing to note here is that we are testing if metacopy is being used

Ah, gotcha; I should've looked more in-depth, yes in that case naming makes sense

neersighted commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -107,3 +107,59 @@ func doesSupportNativeDiff(d string) error {
+// copying up a file from a lower layer unless/until its contents are being
+// modified
+func doesMetacopy(d string) (bool, error) {
+	td, err := os.MkdirTemp(d, "metacopy-check")
```
I was unsure of this myself -- some guidelines on how to run the test harness in a user namespace would be good. As far as I am aware this is merely a fast return but the slower path of actually doing a mount will be valid in a user namespace regardless of what the correct answer here is.

thaJeztah replied:
@AkihiroSuda can you help with answering this one?

AkihiroSuda replied:
When running in userns you have to :
- set `userxattr` mount option (https://github.com/containers/storage/blob/3940c231cbbb1d16bb65bcbdc21fe0f288f6449a/drivers/overlay/check.go#L152-L154)
-  s/trusted/user/  (https://github.com/containers/storage/blob/3940c231cbbb1d16bb65bcbdc21fe0f288f6449a/drivers/overlay/check.go#L176)

neersighted commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -266,6 +280,7 @@ func (d *Driver) Status() [][2]string {
 	return [][2]string{
 		{"Backing Filesystem", backingFs},
 		{"Supports d_type", strconv.FormatBool(d.supportsDType)},
+		{"Using metacopy", strconv.FormatBool(d.usingMetacopy)},
```
See https://github.com/moby/moby/pull/43557#discussion_r868288711

neersighted commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -159,6 +160,18 @@ func Init(home string, options []string, idMap idtools.IdentityMapping) (graphdr
+		} else {
+			logger.Debugf("overlay test mount indicated that metacopy is not being used")
+		}
+	} else {
+		logger.Warnf("overlay test mount did not indicate whether or not metacopy is being used: %v", err)
+		return nil, err
+	}
```
Okay, I'm looking at this again and thinking that we should still bubble any errors up out of `doesMetacopy()` -- specifically, any failure to Mkdir/Chmod/manipulate the filesystem in any way should be bubbled up instead of swallowed by `doesMetacopy()`. 

While yes, it is optional/informational, I don't think it makes sense to silently ignore errors that may indicate an underlying system problem (like being unable to create a temporary directory, or being unable to mount a directory).

neersighted commented:
I've had a go at addressing review feedback -- please let me know if any of the points I pushed back on are problematic.

Regarding this:

> out of interest; this currently only does the detection; should (if supported) this detection also be used by the actual driver?

I don't think there is anything else to be done in the daemon at this time. We don't currently request metacopy (on, or off) in our mounts, so this is merely meant to report the status of metacopy to aid in debugging issues with the daemon/driver in the field. Typically metacopy will be off, as that is the upstream default, but it could be turned on by a custom kernel, a kernel command line, or writing to `/sys/module/overlay/parameters/metacopy`.

In theory we could add configuration to force metacopy on/off and change this check to detect support -- however there is currently no drive to do so. However it does give a good jumping off point if such a feature is ever desirable for some reason.

AkihiroSuda reviewed (dismissed):
https://github.com/moby/moby/pull/43557#discussion_r870125990

neersighted committed:
overlay2: account for UserNS/userxattr in metacopy test

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

neersighted commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -126,6 +127,17 @@ func doesSupportNativeDiff(d string) error {
 func usingMetacopy(d string) (bool, error) {
+	userxattr := false
+	if userns.RunningInUserNS() {
+		needed, err := overlayutils.NeedsUserXAttr(d)
```
Note for reviewers: this results in us doing a mount to check if we need `userxattr` twice -- however, I didn't want to attach a major refactor to this PR.

The cost is negligible as this is only done at driver `Init`. There will be a follow-up PR that restructures this to avoid the duplicate work.

thaJeztah commented:
@neersighted looks like a linter is failing;


```

daemon/graphdriver/overlayutils/overlayutils.go:82:1: exported function `GetOverlayXattr` should have comment or be unexported (golint)
 func GetOverlayXattr(name string) string {
 ^
```

thaJeztah reviewed (approved):
LGTM

only wondering if the two commits should be squashed (if the first commit is broken, they probably should be squashed)

cpuguy83 reviewed (approved):
LGTM
